src/Altinn.App.Core/Features/Correspondence/Models/CorrespondenceRequest.cs

```diff
@@ -0,0 +1,272 @@
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics.CodeAnalysis;
+using Altinn.App.Core.Features.Correspondence.Exceptions;
+using Altinn.App.Core.Models;
+
+namespace Altinn.App.Core.Features.Correspondence.Models;
+
+/// <summary>
+/// Base functionality for correspondence models
+/// </summary>
+public abstract record CorrespondenceBase
+{
+    internal static void AddRequired(MultipartFormDataContent content, string value, string name)
+    {
+        if (string.IsNullOrWhiteSpace(value))
+            throw new CorrespondenceValueException($"Required value is missing: {name}");
+
+        content.Add(new StringContent(value), name);
+    }
+
+    internal static void AddRequired(
+        MultipartFormDataContent content,
+        ReadOnlyMemory<byte> data,
+        string name,
+        string filename
+    )
+    {
+        if (data.IsEmpty)
+            throw new CorrespondenceValueException($"Required value is missing: {name}");
+
+        content.Add(new ReadOnlyMemoryContent(data), name, filename);
+    }
+
+    internal static void AddIfNotNull(MultipartFormDataContent content, string? value, string name)
+    {
+        if (!string.IsNullOrWhiteSpace(value))
+            content.Add(new StringContent(value), name);
+    }
+
+    internal static void AddListItems<T>(
+        MultipartFormDataContent content,
+        IReadOnlyList<T>? items,
+        Func<T, string> valueFactory,
+        Func<int, string> keyFactory
+    )
+    {
+        if (IsEmptyCollection(items))
+            return;
+
+        for (int i = 0; i < items.Count; i++)
+        {
+            string key = keyFactory.Invoke(i);
+            string value = valueFactory.Invoke(items[i]);
+            content.Add(new StringContent(value), key);
+        }
+    }
+
+    internal static void SerializeListItems(
+        MultipartFormDataContent content,
+        IReadOnlyList<CorrespondenceListBase>? items
+    )
+    {
+        if (IsEmptyCollection(items))
+            return;
+
+        for (int i = 0; i < items.Count; i++)
+        {
+            items[i].Serialise(content, i);
+        }
+    }
+
+    internal static void SerializeAttachmentItems(
+        MultipartFormDataContent content,
+        IReadOnlyList<CorrespondenceAttachment>? attachments
+    )
+    {
+        if (IsEmptyCollection(attachments))
+            return;
+
+        // Ensure unique filenames
+        var overrides = new Dictionary<CorrespondenceAttachment, string>();
```

Comment:
Since `CorrespondenceAttachment` is a record and it is being used as a key here, it is doing value equality across all the fields within. I wonder if this is actually a good case for reference equality since it is a class?

Reply:
Good catch!